src/device/mod.rs

```diff
@@ -1484,36 +1494,25 @@ impl Runtime {
             .map(|peers| peers.iter().any(|p| p.public_key == exit_node.public_key))
             .unwrap_or_default();
 
-        self.requested_state.postquantum_wg = None;
-
         if is_meshnet_exit_node {
             if let Some(dns) = &self.entities.dns.lock().await.resolver {
                 self.reconfigure_dns_peer(dns, &dns.get_default_dns_servers())
                     .await?;
             }
         } else if let Some(addr) = exit_node.endpoint {
             if let Some(pq_conf) = &self.features.post_quantum_vpn {
-                telio_log_debug!("Initializing PQ hanshake");
+                self.requested_state.postquantum_keys = None;
 
-                let fetch_keys = Box::pin(wg::pq::fetch_keys(
-                    &self.entities.socket_pool,
+                let entt = telio_pq::Entity::run(
+                    self.event_publishers.post_quantum.clone(),
+                    self.entities.socket_pool.clone(),
                     addr,
-                    &self.requested_state.device_config.private_key,
-                    &exit_node.public_key,
-                )); // The future is large, let's move it onto the heap
-
-                let keys = tokio::time::timeout(
-                    Duration::from_secs(pq_conf.handshake_timeout_s as _),
-                    fetch_keys,
-                )
-                .await
-                .map_err(|_| {
-                    telio_log_warn!("PQ hanshake timeout");
-                    wg::pq::Error::Generic("Fetching PQ keys timeout".into())
-                })??;
+                    self.requested_state.device_config.private_key,
+                    exit_node.public_key,
+                    pq_conf,
+                );
 
-                self.requested_state.postquantum_wg = Some(keys);
-                telio_log_debug!("PQ hanshake finished succesfully");
+                self.entities.postquantum_wg = Some(entt);
```

Comment:
What would happen if `telio_connect_to_exit_node()` is called while `self.entities.postquantum_wg` is already `Some(...)`.

Dont we have to kill/stop etc. tokio task spawned on "previous" entity?

Reply:
It will be killed in the `impl Drop for Entity`, so this case if handled